internal/controller/changetransferpolicy_controller.go

```diff
@@ -155,6 +162,227 @@ func (r *ChangeTransferPolicyReconciler) Reconcile(ctx context.Context, req ctrl
+// calculateHistory calculates the history by getting the first parents on the active branch and using the trailers to reconstruct the history.
+func (r *ChangeTransferPolicyReconciler) calculateHistory(ctx context.Context, ctp *promoterv1alpha1.ChangeTransferPolicy, gitOperations *git.EnvironmentOperations) error {
+	logger := log.FromContext(ctx)
+
+	shaListActive, err := gitOperations.GetRevListFirstParent(ctx, "origin/"+ctp.Spec.ActiveBranch, 5)
+	if err != nil {
+		return fmt.Errorf("failed to get rev-list commit history for branch %q: %w", ctp.Spec.ActiveBranch, err)
+	}
+	logger.V(4).Info("Rev-list history for active branch", "shaList", shaListActive)
+
+	history := make([]promoterv1alpha1.History, 0, len(shaListActive))
+	for _, sha := range shaListActive {
+		historyEntry, shouldInclude, err := r.buildHistoryEntry(ctx, sha, gitOperations)
+		if err != nil {
+			logger.Error(err, "failed to build history entry", "sha", sha)
+			continue
+		}
+
+		if shouldInclude {
+			history = append(history, historyEntry)
+		}
+	}
+
+	ctp.Status.History = history
+	return nil
+}
+
+// buildHistoryEntry creates a single history entry for the given SHA
+func (r *ChangeTransferPolicyReconciler) buildHistoryEntry(ctx context.Context, sha string, gitOperations *git.EnvironmentOperations) (promoterv1alpha1.History, bool, error) {
+	activeTrailers, err := gitOperations.GetTrailers(ctx, sha)
+	if err != nil {
+		return promoterv1alpha1.History{}, false, fmt.Errorf("failed to get trailers for SHA %q: %w", sha, err)
+	}
+
+	// Skip no-op commits
+	if activeTrailers[constants.TrailerNoOp] == "true" {
+		return promoterv1alpha1.History{}, false, nil
+	}
+
+	historyEntry := promoterv1alpha1.History{
+		Proposed:    promoterv1alpha1.CommitBranchStateHistoryProposed{},
+		Active:      promoterv1alpha1.CommitBranchState{},
+		PullRequest: &promoterv1alpha1.PullRequestCommonStatus{},
+	}
+
+	r.populateActiveMetadata(ctx, &historyEntry, sha, gitOperations)
+	r.populateProposedMetadata(ctx, &historyEntry, activeTrailers, gitOperations)
+	r.populatePullRequestMetadata(ctx, &historyEntry, activeTrailers)
+	r.populateCommitStatuses(ctx, &historyEntry, activeTrailers)
+
+	return historyEntry, true, nil
+}
+
+// populateActiveMetadata populates the active metadata for a history entry
+func (r *ChangeTransferPolicyReconciler) populateActiveMetadata(ctx context.Context, h *promoterv1alpha1.History, sha string, gitOperations *git.EnvironmentOperations) {
+	logger := log.FromContext(ctx)
+
+	if activeHydrated, err := gitOperations.GetShaMetadataFromGit(ctx, sha); err != nil {
+		logger.Error(err, "failed to get hydrated commit metadata for active SHA", "activeSha", sha)
+	} else {
+		h.Active.Hydrated = activeHydrated
+		h.Active.Hydrated.Body = removeKnownTrailers(h.Active.Hydrated.Body)
+	}
+
+	if activeDry, err := gitOperations.GetShaMetadataFromFile(ctx, sha); err != nil {
+		logger.Error(err, "failed to get dry commit metadata for active SHA", "activeSha", sha)
```

Comment:
This file should exist even after a manual merge, so we should throw the error if we fail to get it.